Add initial ActivityPub support to starter packs

jfietkau opened the pull request:
Building on #175, this pull request adds ActivityPub support to starter packs. This is not intended to be ready to merge right away, there are several spots in here that I want to discuss. Plus it'll only make sense to actually run in practice once the ActivityPub backfilling from the post-merge discussion in #175 is in place. But we can already start talking about it. And if you want to merge it early, it does gracefully handle missing ActivityPub IDs, so the only hard depencency is the stuff from #175 that's already merged.

About the design of what's here:

I took a "keep it simple" approach under the assumption that there is currently nothing else in Fedidevs that we want to add ActivityPub support to. So the changes here are all contained in `starter_packs/views.py` even though the `wants_activitypub` function, which detects HTTP requests from ActivityPub platforms, is fit for the general case. If we ever do more ActivityPub in various places, we should find a better spot for it. Until then I reckon it makes sense to keep it all close by.

In this implementation, we use content negotiation to serve ActivityPub under the same URL as the HTML version. I've read some Django docs saying that the view is the correct place to do this. There are other options besides content negotiation, such as `Link` HTTP headers or HTML elements. [There is a draft document](https://swicg.github.io/activitypub-html-discovery/) by the SWICG that gives an overview. But again, I believe this is the simplest approach, and also coincidentally the one with the best compatibility in the ecosystem.

Because we serve only a single ActivityPub schema, it makes more sense to construct it directly from a `dict` than to pull in dependencies related to JSON-LD. One last time: keeping it simple.

The schema itself is the one from #pixelfed/starter-kits/issues/1 with some of my own suggested additions. But we should be compatible with it as well as with the ActivityStreams base vocabulary.

I'm going to try to add a couple of line comments to highlight spots where I wasn't 100% sure.

anze3db commented on starter_packs/views.py:
```diff
@@ -407,6 +419,43 @@ def share_starter_pack(request, starter_pack_slug):
     )
 
+    if wants_activitypub(request):
+        author = Account.objects.get(username_at_instance=starter_pack.created_by)
```
I think what you'll need to do is:

```suggestion
        author = Account.objects.get(username_at_instance=starter_pack.created_by.username)
```

created_by isn't nullable and an auth.User should always have a corresponding Account so I think this should work 🤞

anze3db commented on starter_packs/views.py:
```diff
@@ -407,6 +419,43 @@ def share_starter_pack(request, starter_pack_slug):
+        data = {
+            "@context": "https://www.w3.org/ns/activitystreams",
+            "type": "Collection",
+            "id": request.build_absolute_uri("/s/" + starter_pack.slug + "/"),
```
`/s/slug/` is the only path for each starter pack. I would use django's reverse function to get the `/s/slug/` part though:

```suggestion
            "id": request.build_absolute_uri(reverse("share_starter_pack", kwargs={"starter_pack_slug": starter_pack.slug})),
```

anze3db commented on starter_packs/views.py:
```diff
@@ -407,6 +419,43 @@ def share_starter_pack(request, starter_pack_slug):
+            "image": {
+                "type": "Image",
+                "mediaType": "image/png",
+                "url": request.build_absolute_uri("/static/og-starterpack.png"),
```
Yes, sounds good 👍

jfietkau commented on starter_packs/views.py:
```diff
@@ -407,6 +419,43 @@ def share_starter_pack(request, starter_pack_slug):
+            # This should be resolved somewhere else by a background task, so we
+            # return an HTTP 500 error and leave it to the client to try again later.
+            return HttpResponseServerError()
+        items = [account.activitypub_id for account in accounts]
```
I'm gonna try my best to set my implementation biases aside and answer from a perspective that mainly takes ActivityPub peers and API clients into account. I think it'll be useful for us to get a bit more shared clarity on this.

> I'm not very familiar with ActivityPub or JSON-LD but in my experience creating an API endpoint with a list of strings is a bad idea since it's impossible to extend without creating a breaking change.

I don't know if "a list of strings" is a useful way to discuss this approach. In JSON-LD, the ID string and the object are quite literally interchangeable, so any consumer of this data needs to accomodate both versions anyway, including switching back and forth between IDs and inlined objects within the same list. It's almost pointless to discuss which one is more convenient for consumers because they need to be able to handle both, regardless of whether they're using a full JSON-LD parser (Fedify, Iceschrimp) or not (Mastodon, Pixelfed). That is to say, it would not be a breaking change to switch between IDs and inlined objects at a later date if we so choose. In fact, it would be a huge mistake for us to guarantee a particular shape of the data structure here.

I'm a tiny bit tempted to randomly replace one in ten IDs with its inlined object counterpart, just to keep ActivityPub consumers on their toes. But I'm not enough of a troll to do that in production software. 🙃

And regarding extensibility, we have very little room to add anything to the contained objects anyway. Because they're owned by other servers, all that we could theoretically add to them would be their properties as they exist on the source server. It's not that I can't see that being useful (e.g. to efficiently compile a list of avatar URLs directly from a starter pack), but I get the impression that you have a perspective on API extensibility here that doesn't map to ActivityPub.

> Also, I'm not sure how useful this endpoint will be if every consumer has to then also refetch all the related objects. Feels wasteful.

If they want to do anything beyond trivial (e.g. check how many entries are in the list) with the starter pack, they need to fetch every listed object from the source servers anyway, since – setting unintentional inaccuracies caused by caching aside – they can't necessarily trust our list to be non-malicious. Imagine if a starter pack hosted on fedidevs.com were allowed to provide an object representation of `https://example.org/users/abc` and the consuming server (say, Pixelfed) just took that at face value. Any starter pack out there would be able to set your name, bio, avatar, etc. to whatever it wants. Since I started using Mastodon, it has had a couple of emergency security updates related to spots where they forgot to validate inlined objects from federated ActivityStreams data against the source instance.

The origin-based data ownership model is documented in [FEP-fe34](https://codeberg.org/fediverse/fep/src/branch/main/fep/fe34/fep-fe34.md). Not a must-read, just some further info I'm linking in case you're curious.

> I'm already refreshing account data somewhat regularly and stale data hasn't been a big issue so far.

That's good to know! 👍 I'll scratch that concern from my mental to do list.

----

With all this in mind, I'm leaning more and more towards the idea that a separate JSON API for consumers who don't care about ActivityPub may very well make sense. But that's outside the scope of this PR.

jfietkau replied:
I was going to object to this on principle, but I investigated the standards a bit more and I... guess we can do that?

From [ActivityStreams 2.0, section 2.1](https://www.w3.org/TR/activitystreams-core/#jsonld):

> Implementations MAY also use additional properties and values not defined in the JSON-LD @​context with the understanding that any such properties will likely be unsupported and ignored by consuming implementations that use the standard JSON-LD algorithms.

That's kind of hilarious. We'd have to avoid name clashes between our added data and the attribute namespaces of AS `Object` and `Collection`, but apart from that I suppose it'll work, even though it makes me feel queasy. I will note though that you're looking at another backfilling script situation, because we're going to have to go through and at least add the ActivityPub `type` to the accounts table.

Do you mind if I briefly loop someone into the conversation who'll know better than me? Hey @silverpill, no need to read all the context, just if you have a moment: adding arbitrary properties to inlined ActivityStreams objects from other origins, for use by plain JSON API consumers, and relying on AS consumers to ignore them. Where does that land on the "meh / weird / rude / illegal" scale?

silverpill replied:
I think adding your properties to other people's objects is "rude". Someone looking at the fetched object and comparing it to the inlined one might think there is a bug. And some newbie developer will inevitably consume the object without fetching it from the authoritative server...

I would wrap it in another object instead:

```json
{
      "id": "https://fedidevs.example/starter-pack-item",
      "object": "https://example.org/users/username", // not inlined
      "name": "Display Name", // might differ from actual display name
      "foo": "bar" // your additional properties
}
```

jfietkau replied:
Thank you! Your stance matches my gut feeling.

Having a wrapper type for entries would break compatibility with Dan's schema (pixelfed/starter-kits/issues/1), which is AFAIK implemented by Pixelfed and WordPress ActivityPub. So I don't think that's something we want to do here.

We could plop our metadata into a separate property of the root `Collection`, which we do own. Although at that point it becomes a little dubious again whether consumers of AS and those wanting plain JSON should really share a payload or whether we should just split them entirely.

anze3db replied:
Fun times 😆

What's the `type` value supposed to contain and where can we get this data?

jfietkau replied:
The value of `type` can in theory be pretty much whatever, although in the wild I've only seen [these five](https://www.w3.org/TR/activitystreams-vocabulary/#actor-types). It differs per account and can change over time.

Unlike the ID, the `type` is sadly not in the Mastodon API account lookup that we already do. The process to acquire it would be to fetch the ActivityPub ID as a URL [with the correct `Accept` header](https://www.w3.org/TR/activitypub/#retrieving-objects) to get the ActivityPub representation. The `type` field is in there and can be stored as a string.

Or we split the ActivityPub and JSON APIs. In that case we could put Mastodon's `bot` boolean that we already have in the database into the JSON result. I think I'd prefer to move this PR in that direction rather than abuse the ActivityPub objects too much. I'll whip something up for you to look at tomorrow.

anze3db commented:
> But hey, if I patch the requirements for a full author account and for ActivityPub IDs out of the view code, the tests succeed. So that's something.

Great! We can definitively make the test have all the necessary data! I'll take a look at this PR tomorrow, it's getting late here so I need to go to bed now 😅

jfietkau commented:
Okay, a new revision of this PR is in. I have taken a shot at separating the JSON and ActivityPub APIs. There's arguably a little bit of duplication, but while putting this together I got the feeling that the results flow much more naturally than trying to cram a general-use JSON API into the ActivityPub format. Let me know what you think.

I'm not fussy about what exactly should be in the JSON result. The way I did it now was to pick a few handfuls of model properties that seemed to make sense to expose, but I'm fairly dispassionate about the JSON API. I'm probably not going to be among its users. At this stage, this PR is more about the general structure of the code and the API separation.

While I was at it, I added the ability to request a specific format via a GET parameter, so you can do `/s/slug/?format=json` (or `html` or `activitypub`) to look at the differences. It's a bit handier for me too than having to fiddle with the `Accept` header each time, so I wanted to get this done anyway.

I've also updated the tests. They're still missing their setup, but the JSON API has preliminary validation.
